src/Disks/DiskCache.cpp

```diff
@@ -0,0 +1,1026 @@
+        cache_policy(std::move(cache_policy_)), cache_key(cache_key_), cache_base_path(cache_base_path_),
+        current_offset(offset_), end_offset(size_ ? offset_ + size_ : size_),
+        downloader(std::move(downloader_))
+    {
+        ProfileEvents::increment(ProfileEvents::DiskCacheRequestsIn, 1);
+        log = &Poco::Logger::get("CacheableMultipartReadBufferDecorator");
+        ensureFrom();
+    }
+
+    void set(BufferBase::Position ptr, size_t size, size_t offset) override
+    {
+        from->BufferBase::set(ptr, size, offset);
+    }
+
+    BufferBase::Buffer & internalBuffer() override
+    {
+        throw Exception("Method not implemented", ErrorCodes::NOT_IMPLEMENTED);
+    }
+
+    BufferBase::Buffer & buffer() override
+    {
+        return from->buffer();
+    }
+
+    BufferBase::Position & position() override
+    {
+        return from->position();
+    }
+
+    size_t offset() const override
+    {
+        return from->offset();
+    }
+
+    size_t available() const override
+    {
+        return from->available();
+    }
+
+    void swap(BufferBase &) override
+    {
+        throw Exception("Method not implemented", ErrorCodes::NOT_IMPLEMENTED);
+    }
+
+    size_t count() const override
+    {
+        throw Exception("Method not implemented", ErrorCodes::NOT_IMPLEMENTED);
+    }
+
+    bool hasPendingData() const override
+    {
+        return from->hasPendingData();
+    }
+
+    bool isPadded() const override
+    {
+        throw Exception("Method not implemented", ErrorCodes::NOT_IMPLEMENTED);
+    }
+
+    void set(BufferBase::Position ptr, size_t size) override
+    {
+        from->set(ptr, size);
+    }
+
+    bool next() override
+    {
+        bool res = from->next();
+        if (!res)
+        {
+            from.reset(nullptr);
+            ensureFrom();
+            if (from)
+                res = from->next();
+        }
+
+        if (res)
+        {
+            auto buffer = from->buffer();
+            size_t size = buffer.size();
+            ProfileEvents::increment(ProfileEvents::DiskCacheBytesTotal, size);
+
+            switch (current_type)
+            {
+                case DiskCachePolicy::CachePart::CachePartType::CACHED:
+                    ProfileEvents::increment(ProfileEvents::DiskCacheBytesCached, size);
+                    break;
+                case DiskCachePolicy::CachePart::CachePartType::ABSENT:
+                    ProfileEvents::increment(ProfileEvents::DiskCacheBytesRemote, size);
+                    ProfileEvents::increment(ProfileEvents::S3ReadBytes, size);
+                    break;
+                case DiskCachePolicy::CachePart::CachePartType::ABSENT_NO_CACHE:
+                    ProfileEvents::increment(ProfileEvents::DiskCacheBytesRemote, size);
+                    ProfileEvents::increment(ProfileEvents::DiskCacheBytesRemoteSkipCache, size);
+                    ProfileEvents::increment(ProfileEvents::S3ReadBytes, size);
```

Comment:
btw `S3ReadBytes`, which is incremented in `readBufferFromS3` is not in fact updated for prefetches (part of asynchronous reads feature), because reading is done in a  threadpool and `increment` method counts on that: `DB::CurrentThread::getProfileEvents().increment(...)`.
(when async reads setting was put to 1 by default. some tests which checked how ProfileEvents::S3ReadBytes is updated -- got broken, so we decided to remove those checks at least for now..)

May be makes sense to use profile event which are on `AsynchronousReadIndirectBufferFromRemoreFS` and the same non-async version - `ReadIndirectBufferFromRemoteFS`.

Reply:
In top-level buffer is no info what part of data was cached and what part was donwloaded right now, so this events should be in cache level or deeper.

Reply:
but `S3ReadBytes` is also updated in `ReadBufferFromS3`, does it make sense to update here too?

Reply:
In ReadBufferFromS3 this updated only when cache is disabled.
It may be request with and without cache at same time, for example cache now is not applied for S3 table function - currently cache is a part of DiskS3, but S3 table function makes direct request to remote storage without DiskS3.
I agree that this is confused. May be better to make one download method and use it in both places.

Reply:
> In ReadBufferFromS3 this updated only when cache is disabled.

Yes, but in this line it get under case `ABSENT_NO_CACHE`, so looks like in this case it will read via `ReadBufferFromS3` and one more time update `S3ReadBytes`, isn't it?

Reply:
No, ABSENT_NO_CACHE is "emergency" option for max attempts reach or aviodinig deadlocks. For ReadBufferFromS3 this variant also "reading from cache", but with skipping write readed data in local cache.